rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp

```diff
@@ -0,0 +1,220 @@
+    Node(node_name, use_intra_process_comms)
+  {
+    setup_state_machine();
+  }
+
+  LIFECYCLE_EXPORT
+  ~LifecycleNode(){}
+
+  /**
+   * @brief get the default state machine
+   * as defined on design.ros.org
+   */
+  LIFECYCLE_EXPORT
+  virtual void
+  setup_state_machine()
+  {
+    state_machine_ = rcl_get_default_state_machine();
+  }
+
+  /**
+   * @brief same API for creating publisher as regular Node
+   */
+  template<typename MessageT, typename Alloc = std::allocator<void>>
+  std::shared_ptr<rclcpp::publisher::LifecyclePublisher<MessageT, Alloc>>
+  create_publisher(
+      const std::string & topic_name,
+      const rmw_qos_profile_t & qos_profile = rmw_qos_profile_default,
+      std::shared_ptr<Alloc> allocator = nullptr)
+  {
+    // create regular publisher in rclcpp::Node
+    auto pub =  rclcpp::node::Node::create_publisher<MessageT, Alloc,
+      rclcpp::publisher::LifecyclePublisher<MessageT, Alloc>>(
+        topic_name, qos_profile, allocator);
+
+    // keep weak handle for this publisher to enable/disable afterwards
+    weak_pub_ = pub;
+    return pub;
+  }
+
+  LIFECYCLE_EXPORT
+  void
+  print_state_machine()
+  {
+    printf("current state is %s\n", state_machine_.current_state.label);
+    rcl_print_transition_map(&state_machine_.transition_map);
+  }
+
+  LIFECYCLE_EXPORT
+  virtual bool
+  on_configure()
+  {
+    // check on every function whether we are in the correct state
+    /*if (!rcl_is_valid_transition(&state_machine_, &rcl_state_inactive))
+    {
+      // if not, go to error state
+      state_machine_.current_state = rcl_state_error;
+      return false;
+    }*/
+
+    // Placeholder print for all configuring work to be done
+    // with each pub/sub/srv/client
+    printf("I am doing some important configuration work\n");
+
+    // work was done correctly, so change the current state
+    state_machine_.current_state = rcl_state_inactive;
+    return true;
+  }
+
+  LIFECYCLE_EXPORT
+  virtual bool
+  on_clean_up()
+  {
+    return true;
+  }
+
+  LIFECYCLE_EXPORT
+  virtual bool
+  on_shutdown()
+  {
+    return true;
+  }
+
+  LIFECYCLE_EXPORT
+  virtual bool
+  on_activate()
+  {
+    /*if (!rcl_is_valid_transition(&state_machine_, &rcl_state_active))
+    {
+      fprintf(stderr, "Unable to change from  current state %s from transition start %s\n", state_machine_.current_state.label, rcl_state_active.label);
+      state_machine_.current_state = rcl_state_error;
+      return false;
+    }*/
+    if (weak_pub_.expired())
+    {
+      // Someone externally destroyed the publisher handle
+      fprintf(stderr, "I have no publisher handle\n");
+      state_machine_.current_state = rcl_state_error;
+      return false;
+    }
+
+    // Placeholder print for all configuring work to be done
+    // with each pub/sub/srv/client
+    printf("I am doing a lot of activation work\n");
+    // TODO: does a return value make sense here?
+    weak_pub_.lock()->on_activate();
```

Comment:
Race condition: weak_pub_ may have been deleted between the check on `weak_pub_.expired()` and this statement. Use something like:

    auto pub = weak_pub_.lock();
    if (pub) {
      pub->on_activate();
    }
    else {
      [Your error stuff]
    }